content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md

```diff
@@ -0,0 +1,146 @@
+
+## Using Transition Rules
+
+[Transition Rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#transition-rules) make it possible to compare the new state against the old state of a resource in validation rules.  Any validation rule that uses 'oldSelf' is a transition rule.  Transition rules are only evaluated when an old value and new value exist.
+
+Transition rule examples:
+
+| Transition Rule | Purpose |
+| - | - |
+| `self == oldSelf` | Validate that a required field is immutable once it is set. This does not prevent an optional field from transitioning from unset to set or set to unset.|
+| on parent of field: `has(self.field) == has(oldSelf.field)`<br>on field:  `self == oldSelf` | Validate that an optional field never changes after the resource is created.|
+| `self.all(x, x in oldSelf)` | Validate that a set is append-only.|
+| `self >= oldSelf` | Validate that a number is monotonically increasing.|
+
+
+## Using the Functions Libraries
+
+Validation rules have access to a couple different function libraries:
+
+* CEL standard functions, defined in the [list of standard definitions](https://github.com/google/cel-spec/blob/v0.7.0/doc/langdef.md#list-of-standard-definitions)
+* CEL standard [macros](https://github.com/google/cel-spec/blob/v0.7.0/doc/langdef.md#macros)
+* CEL [extended string function library](https://pkg.go.dev/github.com/google/cel-go@v0.11.2/ext#Strings)
+* Kubernetes [CEL extension library](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#pkg-functions) which includes supplemental functions for [lists](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#pkg-functions), [regex](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#Regex), and [URLs](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#Regex).
+
+Examples of function libraries in use:
+
+| Validation Rule | Purpose |
+| - | - |
+| `!(self.getDayOfWeek() in [0, 6])` | Validate that a date is not a Sunday or Saturday. |
+| `isUrl(self) && url(self).getHostname() in [a.example.com', 'b.example.com']` | Validate that a URL has an allowed hostname. |
+| `self.map(x, x.weight).sum() == 1` | Validate that the weights of a list of objects sum to 1. |
+| `int(self.find('^[0-9]*')) < 100` | Validate that a string starts with a number less than 100. |
+| `self.isSorted()` | Validates that a list is sorted. |
+
+## Resource Limits
+
+To prevent CEL evaluation from consuming excessive compute resources, validation rules impose some limits. These limits are based on CEL "cost units", a platform and machine independent measure of execution cost. As a result, the limits are the same regardless of where they are enforced.
```

Comment:
```suggestion
To prevent CEL evaluation from consuming excessive compute resources, validation rules impose some limits. These limits are based on CEL _cost units_, a platform and machine independent measure of execution cost. As a result, the limits are the same regardless of where they are enforced.
```